### PR TITLE
AUTH-367 add scope claim to Token class

### PR DESCRIPTION
**Pull request by ahmadfaridulazmi**
JIRA: https://volcanicuk.atlassian.net/browse/AUTH-367

added instance variable `scope` to `Token` class.
added `privileges_for_scope` method to `Subject` class.
added options `privilege_ids` to `Identity` creation.

### CODE REVIEW

File: lib/volcanic/authenticator/v1/subject.rb [hunk 1]
```diff
@@ -23,15 +23,27 @@ def privileges_for(subject, service)
             exception,
             "#{path}/privileges?filter[subject]=#{subject}&filter[service]=#{service}"
           )
-          return [] if response.blank?
+          privilege_parser(response)
+        end
+
+        # similar to +privileges_for+, accept it
+        # search privileges by token scope
+        def privileges_for_scope(token_base64, service)
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/subject.rb:31]: `privileges_for_scope` is a confusing name.... we're not getting the privileges related to the scope, but to the specified token. A better name would be `privileges_for_token`.

Given that this has nothing to do with the subject, the method is probably in the wrong place... It is probably code that should be in the `Token` class

File: lib/volcanic/authenticator/v1/subject.rb [hunk 2]
```diff
@@ -23,15 +23,27 @@ def privileges_for(subject, service)
...
+          response = perform_get_and_parse(exception, "#{path}/privileges/identity&filter[service]=#{service}", token_base64)
+          privilege_parser(response)
+        end
+
+        private
+
+        def privilege_parser(res)
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/subject.rb:38]: This code isn't about subjects... it is about privileges...

It should probably be in the `Privilege` class

File: lib/volcanic/authenticator/v1/subject.rb (hunk 1 above)
Comment (ahmadfaridulazmi) [reply lib/volcanic/authenticator/v1/subject.rb:31]: both method `privileges_for_scope` and `privileges_for` is calling the same resource path which is `'api/v1/subject'`. That is why thought of applying it at `subject` class.

**Review by danleyden — APPROVED** [review]
Needs version bumping - I think to 1.5.0

it is a `minor` release - adding functionality, not breaking

**Pull request merged**